src/main/kotlin/org/rust/ide/utils/import/importUtils.kt

```diff
@@ -129,10 +124,39 @@ private fun RsUseItem.tryGroupWith(
     val newUsePath = parentPath.joinToString("::", postfix = "::") +
         (importingNames + newImportingName).joinToString(", ", "{", "}")
     val newUseSpeck = psiFactory.createUseSpeck(newUsePath)
+
+    val isUseSpeckSorted = useSpeck?.useGroup?.useSpeckList?.isSortedWith(COMPARATOR_FOR_SPECKS_IN_USE_GROUP) ?: true
+    if (isUseSpeckSorted) {
+        newUseSpeck.useGroup?.sortUseSpecks()
+    }
+
     useSpeck?.replace(newUseSpeck)
     return true
 }
 
+private fun tryInsertUseItemAtCorrectLocation(mod: RsItemsOwner, useItem: RsUseItem): Boolean {
+    val newline = RsPsiFactory(mod.project).createNewline()
+    val uses = mod.childrenOfType<RsUseItem>().map(::UseItemWrapper)
+    if (uses.isEmpty()) {
+        val anchor = mod.childrenOfType<RsExternCrateItem>().lastOrNull() ?: return false
+        mod.addBefore(newline, mod.addAfter(useItem, anchor))
+        return true
+    }
+
+    val useWrapper = UseItemWrapper(useItem)
+    val (less, greater) = uses.partition { it < useWrapper }
+    val anchorBefore = less.lastOrNull()
+    val anchorAfter = greater.firstOrNull()
+    when {
+        anchorBefore != null ->
+            mod.addBefore(newline, mod.addAfter(useItem, anchorBefore.useItem))
+        anchorAfter != null ->
+            mod.addAfter(newline, mod.addBefore(useItem, anchorAfter.useItem))
```

Comment:
```suggestion
        anchorBefore != null -> {
            val addedItem = mod.addAfter(useItem, anchorBefore.useItem)
            mod.addBefore(newline, addedItem)It should significantly simplify reading. Otherwise, the code makes me think more about execution order
        }
        anchorAfter != null -> {
            val addedItem = mod.addBefore(useItem, anchorAfter.useItem)
            mod.addAfter(newline, addedItem)
        }
```
It should significantly simplify reading. Otherwise, the code makes me think more about execution order